src/t_set.c

```diff
@@ -1130,10 +1130,15 @@ void sunionDiffGenericCommand(client *c, robj **setkeys, int setnum,
          * This way we perform at max N*M operations, where N is the size of
          * the first set, and M the number of sets. */
         si = setTypeInitIterator(sets[0]);
+        int sameset = 0;
         while((ele = setTypeNextObject(si)) != NULL) {
             for (j = 1; j < setnum; j++) {
                 if (!sets[j]) continue; /* no key is an empty set. */
-                if (sets[j] == sets[0]) break; /* same set! */
+                if (sets[j] == sets[0]) {
+                    /* same set! */
+                    sameset = 1;
+                    break;
+                }
```

Comment:
you mean the user provided the same input key twice (to the SDIFF* command).
not sure we should optimize for this case, but if we do, any reason not so handle sdiff_algo==2 too?

one nicer way to doe that IMHO is to do the `sameset` detection in the loop that does `lookupKeyRead`
and then in the if-else chain that handles the `op`, add a first one that does
```
    } else if (op == SET_OP_DIFF && sameset) {
        /* nothing. */
    } else if (op == SET_OP_DIFF && sets[0] && diff_algo == 1) {
    ....
```

Reply:
I mentioned in my comment above that zset has optimizations for this.
We just need to check when choosing the diff algorithm as in zset,would that be better?

Reply:
sorry, i composed my comment before you posted that comment and didn't get to publish it.
didn't notice your comment till now.
anyway, feel free to edit your PR with what you think is best, and then i'll comment again.